Merge Task Server & Build Server

Jiaaming opened the pull request:
https://github.com/microsoft/build-server-for-gradle/pull/162
FYI:
<img width="728" alt="Screenshot 2024-07-10 at 14 52 03" src="https://github.com/microsoft/vscode-gradle/assets/71547730/2a590fed-0c48-4af4-9f78-12e62cb8258d">

<img width="604" alt="Screenshot 2024-07-10 at 14 52 42" src="https://github.com/microsoft/vscode-gradle/assets/71547730/f2737ebf-2fd6-4cfb-8121-8ef54b87afbd">

Commit:
merge

Fix buildJars task (#1481)

chore(deps): bump protobufjs from 7.2.4 to 7.2.6 in /npm-package (#1483)

Bumps [protobufjs](https://github.com/protobufjs/protobuf.js) from 7.2.4 to 7.2.6.
- [Release notes](https://github.com/protobufjs/protobuf.js/releases)
- [Changelog](https://github.com/protobufjs/protobuf.js/blob/master/CHANGELOG.md)
- [Commits](https://github.com/protobufjs/protobuf.js/compare/protobufjs-v7.2.4...protobufjs-v7.2.6)

---
updated-dependencies:
- dependency-name: protobufjs
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jinbo Wang <jinbwan@microsoft.com>

chore(deps-dev): bump protobufjs from 7.2.4 to 7.2.6 in /extension (#1482)

Bumps [protobufjs](https://github.com/protobufjs/protobuf.js) from 7.2.4 to 7.2.6.
- [Release notes](https://github.com/protobufjs/protobuf.js/releases)
- [Changelog](https://github.com/protobufjs/protobuf.js/blob/master/CHANGELOG.md)
- [Commits](https://github.com/protobufjs/protobuf.js/compare/protobufjs-v7.2.4...protobufjs-v7.2.6)

---
updated-dependencies:
- dependency-name: protobufjs
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jinbo Wang <jinbwan@microsoft.com>

chores: Add build server debug config (#1486)

build: Update code sign task to v5 (#1488)

refactor: Move Gradle Daemons implementation from Java into Typescript (#1489)

Fix: Append the log message to build output if it's from compilation report. (#1490)

- Store the id of the compilation task in an LRU cache.
- Append the log message to the build output if it's from the compilation task.

Signed-off-by: Sheng Chen <sheche@microsoft.com>

fix: set GradleExecution jdk use java.import.gradle.java.home (#1491)

Refresh the output folders after build tasks (#1493)

Signed-off-by: Sheng Chen <sheche@microsoft.com>

build: Update telemetry wrapper to 0.14.0 (#1495)

chore(deps-dev): bump braces from 3.0.2 to 3.0.3 in /extension (#1496)

Bumps [braces](https://github.com/micromatch/braces) from 3.0.2 to 3.0.3.
- [Changelog](https://github.com/micromatch/braces/blob/master/CHANGELOG.md)
- [Commits](https://github.com/micromatch/braces/compare/3.0.2...3.0.3)

---
updated-dependencies:
- dependency-name: braces
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Extract api scan to a separate pipeline (#1498)

build: Prepare for 3.14.0 (#1499)

chore(deps-dev): bump braces from 3.0.2 to 3.0.3 in /npm-package (#1500)

Bumps [braces](https://github.com/micromatch/braces) from 3.0.2 to 3.0.3.
- [Changelog](https://github.com/micromatch/braces/blob/master/CHANGELOG.md)
- [Commits](https://github.com/micromatch/braces/compare/3.0.2...3.0.3)

---
updated-dependencies:
- dependency-name: braces
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

gradle server muti-thread

fix- Fix the importer version and prepare for 3.14.1 (#1501)

build - Update target platform (#1502)

init named pipe

impl muti-thread and named pipe for buildServer

gradle server muti-thread

fix- Fix the importer version and prepare for 3.14.1 (#1501)

build - Update target platform (#1502)

init named pipe

test build Server named pipe connection

test build Server named pipe connection

update build/initialize sender

finish forward message demo

finish merge and named pipe

implement polling solution

Only send request after initialization (#1503)

fix - Add java 22 to compatibility matrix (#1505)

Introduce gson to simplify the object parsing (#1509)

Signed-off-by: Sheng Chen <sheche@microsoft.com>

fix bug: buildServerHandler

Commit:
merge server

Commit:
merge both server

Commit:
fix lint

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterNamedPipeStream.java:
```diff
@@ -0,0 +1,247 @@
+ *
+ */
+
+public class ImporterNamedPipeStream {
```
Rename to NamedPipeStream?

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -32,16 +28,14 @@ public class ImporterPlugin extends Plugin {
-
     private static String bundleVersion = "";
 
+	private static String bundleDirectory;
```
there are some formatter(indentation) issues in this file

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterNamedPipeStream.java:
```diff
@@ -0,0 +1,247 @@
+            path = "\\\\.\\pipe\\" + randomSuffix + "-" + type + "-sock";
+        } else {
+			//save pipe file to extension folder
+            path = new File(this.bundleDir).getParent() +'/'+ randomSuffix + "-" + type + ".sock";
```
Why we need to save the pipe file to here?

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/GradleBuildServerProjectImporter.java:
```diff
@@ -8,6 +8,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.concurrent.CompletionException;
```
Unused import

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -75,6 +69,12 @@ public static DigestStore getDigestStore() {
         return instance.digestStore;
     }
 
+    /**
```
Conflict not resolved correctly.

Commit:
fix generateRandomPipeName

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -97,82 +89,29 @@ public static BuildServerConnection getBuildServerConnection(IPath rootPath, boo
-
-        return ":"; // Linux or Mac
-    }
+		try {
```
Please use space for indentation

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -97,82 +89,29 @@ public static BuildServerConnection getBuildServerConnection(IPath rootPath, boo
+
+			instance.buildServers.put(rootPath, Pair.of(server, client));
+			return server;
+    	} catch (Exception e) {
```
Please avoid using general Exception

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -97,82 +89,29 @@ public static BuildServerConnection getBuildServerConnection(IPath rootPath, boo
+			instance.buildServers.put(rootPath, Pair.of(server, client));
+			return server;
+    	} catch (Exception e) {
+			e.printStackTrace();
```
This is only for debug purpose, remove it when you want to go production.

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,245 @@
+        if (System.getProperty("os.name").startsWith("Windows")) {
+            return Paths.get("\\\\.\\pipe\\", generateRandomHex(16) + "-sock").toString();
+        }
+		String tmpdir = System.getProperty("java.io.tmpdir");
```
Handle XDG_RUNTIME_DIR: https://github.com/microsoft/vscode-languageserver-node/blob/6d2a392547ea60db1723e027b4b55d268aa282ea/jsonrpc/src/node/main.ts#L183C41-L183C57

jdneo commented on extension/src/bs/BuildServerHandler.ts:
```diff
@@ -0,0 +1,28 @@
+import * as rpc from "vscode-jsonrpc/node";
+import { generateRandomPipeName } from "../util/generateRandomPipeName";
+
+export class BuildServerHandler {
```
Is it possible to encapsulate to one class?

jdneo commented on extension/src/util/generateRandomPipeName.ts:
```diff
@@ -0,0 +1,30 @@
+    ["darwin", 103],
+]);
+
+export function generateRandomPipeName(): string {
```
It should be possible to directly import this util: https://github.com/redhat-developer/vscode-java/blob/38c8b582b40db9644696924f899c73a3251563f4/src/javaServerStarter.ts#L9

Jiaaming replied:
Currently use local implementation and marked as TODO, should be replaced after upgrade `"vscode-languageclient": "7.0.0"` to higher version

Commit:
refine named pipe repost design

Jiaaming commented on extension/src/server/GradleServer.ts:
```diff
@@ -29,27 +31,33 @@ export class GradleServer {
+            const javaExecPath = getRedHatJavaExecutablePath();
 
-            this.logger.debug("Starting server");
+            //Get the Java executable used by JDT.LS, which will be higher than JDK 17.
+            if (!javaExecPath) {
+                await vscode.window.showErrorMessage("No Red Hat Java Extension Pack Found");
+                return;
+            }
+            const serverPipeName = this.messageForwardHandler.getBuildServerPipeName();
+            const args = [String(this.gradleServerPort), serverPipeName, bundleDirectory, javaExecPath];
```
https://github.com/microsoft/vscode-gradle/blob/144c86574922c3a7ff8395e18f041793b8127f15/extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java#L146-L156
These part should be equal with how we get it in Java. Maybe need to add more code to get a Java executable path that higher than JDK17.

jdneo replied:
Two conditions here:

1. If user installs `redhat.java`
    - Check the embedded jre
    - Check setting `java.jdt.ls.java.home` & `java.home`

2. If user does not installs `redhat.java`, in this case, the build server will not be invoked. Then we can use the previous jdk search logic to find a jdk, and launch the server. In the merged server, you can check the version of jdk used to launch itself, and only start the build server when it's higher than 17

Commit:
refine jdk search logic

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,247 @@
+
+            // Need to retry until the pipeName was sent and pipe is created by Extension side
+            boolean connected = false;
+            while (!connected) {
```
Should set something like MAXIMUM_RETIRES to avoid dead loop

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,247 @@
+                    }
+                    connected = true;
+                } catch (IOException e) {
+                    try {
+                        Thread.sleep(1000);
+                    } catch (InterruptedException ie) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException("Thread interrupted while trying to connect to named pipe", ie);
```
Wrap to a function

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,247 @@
+        return hexString.toString();
+    }
+
+    public String generateRandomPipeName() {
```
So far this is not used out-side of this file, please mark it as private first.

jdneo commented on extension/src/bs/MessageForwardHandler.ts:
```diff
@@ -0,0 +1,75 @@
+export const GET_IMPORTER_PIPE_NAME = "gradle.getImporterPipeName";
+
+export class MessageForwardHandler {
+    private buildServerConnection: rpc.MessageConnection | null = null;
```
Name it as MessageProxy, containing two class instance `BuildServerConnection` & `JdtlsImporterConnection`

jdneo commented on gradle-server/build.gradle:
```diff
@@ -12,6 +12,8 @@ java {
 
 dependencies {
   implementation project(":gradle-plugin-api")
+  implementation 'ch.epfl.scala:bsp4j:2.1.0-M4'
```
Is this required?

jdneo commented on extension/src/server/GradleServer.ts:
```diff
@@ -80,6 +94,11 @@ export class GradleServer {
     }
 
+    public async getBundleDirectory(): Promise<string> {
+        const extensionPath = await vscode.commands.executeCommand<string>(GET_EXTENSION_PATH);
```
At line 33, the extension context is already a member of this class, there is no need to setup a command for this purpose

jdneo commented on extension/package.json:
```diff
@@ -131,6 +131,14 @@
+        "title": "Get Build Server Port"
+      },
+      {
+        "command": "gradle.getBundlePath",
```
Where is this command used?

jdneo commented on gradle-server/src/main/java/com/github/badsyntax/gradle/GradleServer.java:
```diff
@@ -54,12 +56,46 @@ private void blockUntilShutdown() throws InterruptedException {
+		Map<String, String> params = parseArgs(args);
+		int gradleServerPort = Integer.parseInt(params.get("port"));
+		String buildServerPipeName = params.get("pipeName");
+		String bundleDirectory = params.get("bundleDir");
```
Check null for these params as well.

Commit:
update messageProxy, add error handling for task server, fix bugs

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -55,35 +56,47 @@ public OutputStream getOutputStream() throws IOException {
+                    attempts++;
                 }
             }
+            if (!connected) {
```
`connected` and `attempts` are redundant. You can achieve the same purpose to check `attempts`

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -55,35 +56,47 @@ public OutputStream getOutputStream() throws IOException {
+            }
+        }
+
+        private void handleConnectionFailure(IOException e, int attempts) {
```
rename: handleConnectionFailure -> sleep

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -202,6 +215,7 @@ public InputStream getInputStream() throws IOException {
     }
 
     public OutputStream getOutputStream() throws IOException {
+
```
nit: this empty line is unnecessary

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    private StreamProvider createProvider() {
+
```
nit: this empty line is unnecessary

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,261 @@
+        }
+    }
+
+    public NamedPipeStream() {}
```
Unnecessary constructor.

Commit:
fix nit

Commit:
fix nit

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/GradleBuildServerProjectImporter.java:
```diff
@@ -61,7 +61,6 @@ public boolean applies(IProgressMonitor monitor) throws OperationCanceledExcepti
             return false;
         }
 
-
```
After the change, build server will not support root folder anymore.

At least, false need to be returned here when current workspace contains multiple roots.

Check: `JavaLanguageServerPlugin.getPreferencesManager().getPreferences().getRootPaths()`

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -101,32 +96,13 @@ public static BuildServerConnection getBuildServerConnection(IPath rootPath, boo
         if (!createIfMissing) {
             return null;
         }
-
```
The paramenter `IPath rootPath` of this method is not used anymore

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -139,40 +115,7 @@ public static BuildServerConnection getBuildServerConnection(IPath rootPath, boo
-        Optional<String> command = ProcessHandle.current().info().command();
-        if (command.isPresent()) {
-            return command.get();
+            "Failed to start build server.", e));
```
indentation

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,255 @@
+            // Need to retry until the pipeName was sent and pipe is created by Extension side
+            while (attempts < MAX_ATTEMPTS) {
+                try {
+                    attemptConnection(pipeFile);
```
no break when connection established successfully, is that ok?

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,255 @@
+        private OutputStream output;
+
+        public PipeStreamProvider() {
+            initializeNamedPipe();
```
Establish connection with retry logic in a constructor is too heavy

jdneo commented on extension/src/bs/BuildServerConnector.ts:
```diff
@@ -0,0 +1,33 @@
+import * as rpc from "vscode-jsonrpc/node";
+import { generateRandomPipeName } from "../util/generateRandomPipeName";
+
+export class BuildServerConnector {
```
Add document for this type

jdneo commented on extension/src/bs/JdtlsImporterConnector.ts:
```diff
@@ -0,0 +1,52 @@
+
+export const GET_IMPORTER_PIPE_NAME = "gradle.getImporterPipeName";
+
+export class JdtlsImporterConnector {
```
Add document for this type

jdneo commented on extension/src/bs/JdtlsImporterConnector.ts:
```diff
@@ -0,0 +1,52 @@
+        return new Promise((resolve) => {
+            this.context.subscriptions.push(
+                vscode.commands.registerCommand(GET_IMPORTER_PIPE_NAME, (pipeName: string) => {
+                    this.importerPipePath = path.resolve(pipeName);
```
This binds the command to this obj, which looks a little bit wired.

I would suggest changing the method to emit an event, And resolve the problem when that event comes.

jdneo commented on extension/src/bs/JdtlsImporterConnector.ts:
```diff
@@ -0,0 +1,52 @@
+        this.importerPipeServer.listen(this.importerPipePath);
+    }
+
+    private setupMessageForwarding(buildServerConnection: rpc.MessageConnection | null): void {
```
This makes the import connector controls the behavior of buildServerConnection, which is not ideal. Because buildServerConnection is not member of importer connector

Commit:
Merge remote-tracking branch 'origin' into merge-with-named-pipe

Commit:
fix comments

Jiaaming commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -102,31 +97,18 @@ public static BuildServerConnection getBuildServerConnection(IPath rootPath, boo
-        command.add(javaExecutablePath);
-        if (Boolean.parseBoolean(System.getenv("DEBUG_GRADLE_BUILD_SERVER"))) {
-            command.add("-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=8989");
+        if (instance.buildServers.size() > 0) {
+            throw new CoreException(new Status(IStatus.ERROR, PLUGIN_ID,
+                    "Not support multiple workspaces."));
```
logic here: should only have one buildServerconnection, one rootPath. if `instance.buildServers.get(rootPath) == null` and `buildServers.size()>0`, meaning that somehow errorly appear another rootPath.

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java:
```diff
@@ -2,9 +2,7 @@
-import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.List;
+
```
nit: unnecessary empty line

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,254 @@
+/**
+ * A class to create a named pipe stream for the importer to communicate with the extension.
+ */
+public class NamedPipeStream {
```
The members order are messy here.

Jiaaming commented on extension/src/server/GradleServer.ts:
```diff
@@ -29,25 +31,38 @@ export class GradleServer {
-            const args = [String(this.port)];
-
-            this.logger.debug("Starting server");
+            let javaExecPath: string | null = null;
+            //Get the Java executable, which will be higher than JDK 17.
+            if (redHatJavaInstalled()) {
+                javaExecPath = getRedHatJavaExecutablePath() || (await getJavaExecutablePath());
+                if (javaExecPath === null) {
+                    await vscode.window.showErrorMessage(NO_JAVA_EXECUTABLE);
+                }
+            }
```
logic here: 
if redHat.java installed: 
- Find embeded jre --(if not found)--> try other way to find java executable with `jdk>=17`
- if still not found, **showErrorMessage**

If not installed:
- meaning that `build server`&`importer` won't be invoked, but we still need to move forward and start MergeServer

Merge Server will check if java executable exist and decide whether start `build server` or not.

jdneo commented on extension/src/bs/BuildServerConnector.ts:
```diff
@@ -0,0 +1,37 @@
+        this.setupServer();
+    }
+
+    private setupServer(): void {
```
nit: Consider to let proxy call this explicitly

jdneo commented on extension/src/server/GradleServer.ts:
```diff
@@ -29,25 +31,38 @@ export class GradleServer {
+            let javaExecPath: string | null = null;
+            //Get the Java executable, which will be higher than JDK 17.
+            if (redHatJavaInstalled()) {
+                javaExecPath = getRedHatJavaExecutablePath() || (await getJavaExecutablePath());
```
Check https://github.com/redhat-developer/vscode-java/blob/38c8b582b40db9644696924f899c73a3251563f4/src/requirements.ts#L36

Commit:
expand jdk search logic and fix comments

Jiaaming commented on extension/src/util/config.ts:
```diff
@@ -26,6 +35,72 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+    const REQUIRED_JDK_VERSION = 17;
+
+    // search from jdt.ls.java.home, java.home
+    const javaHomeGetters = [getJdtlsConfigJavaHome, getConfigJavaHome];
+    let javaHome: string | undefined = undefined;
+    let javaVersion = 0;
+
+    for (const getJavaHome of javaHomeGetters) {
+        javaHome = getJavaHome() || undefined;
+        if (javaHome) {
+            javaVersion = await getMajorVersion(javaHome);
+            // Ensure the Java version is greater than 17
+            if (javaVersion >= REQUIRED_JDK_VERSION) {
+                const javaExecPath = path.join(javaHome, "bin", JAVA_FILENAME);
+                if (fs.existsSync(javaExecPath)) {
+                    return javaExecPath;
+                }
+            }
+        }
+    }
```
This part follow https://github.com/redhat-developer/vscode-java/blob/38c8b582b40db9644696924f899c73a3251563f4/src/requirements.ts#L40-L71,
they search from` jdt.ls.java.home` and  `java.home`. I remove those vscode notifications/messages that sent to users.

Jiaaming commented on extension/src/util/config.ts:
```diff
@@ -26,6 +35,72 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+    }
+
+    // search valid JDKs from env.JAVA_HOME, env.PATH, SDKMAN, jEnv, jabba, Common directories
+    const javaRuntimes = await listJdks();
+    sortJdksByVersion(javaRuntimes);
+
+    const validJdks = javaRuntimes.filter((r) => r.version!.major >= REQUIRED_JDK_VERSION);
+    if (validJdks.length > 0) {
+        sortJdksBySource(validJdks);
+        javaHome = validJdks[0].homedir;
+        javaVersion = validJdks[0].version!.major;
+    }
```
This part follow https://github.com/redhat-developer/vscode-java/blob/38c8b582b40db9644696924f899c73a3251563f4/src/requirements.ts#L74-L100.
Since all this function happens after user already got `redHat.java` install.

Commit:
fix typo

Jiaaming commented on extension/src/util/config.ts:
```diff
@@ -26,6 +35,72 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+    }
+
+    //search java.configuration.runtimes if still not found
+    if (!javaHome) {
+        javaHome = await findDefaultRuntimeFromSettings();
+        javaVersion = await getMajorVersion(javaHome);
+    }
+    if (javaHome && javaVersion >= REQUIRED_JDK_VERSION) {
+        return path.join(javaHome, "bin", JAVA_FILENAME);
+    }
+    return undefined;
```
Follow https://github.com/redhat-developer/vscode-java/blob/38c8b582b40db9644696924f899c73a3251563f4/src/requirements.ts#L101-L103

Commit:
fix typo

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,254 @@
+        return getSelectedStream().getOutputStream();
+    }
+
+    private void sendImporterPipeName(String pipeName) {
```
This should be a member of `PipeStreamProvider`

jdneo commented on extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/NamedPipeStream.java:
```diff
@@ -0,0 +1,254 @@
+        return hexString.toString();
+    }
+
+    private String generateRandomPipeName() {
```
Please move them into the actual class where they are called

jdneo commented on extension/src/bs/MessageProxy.ts:
```diff
@@ -0,0 +1,46 @@
+import * as vscode from "vscode";
+import * as rpc from "vscode-jsonrpc/node";
+
+export class MessageProxy {
```
document:

1. what this class does
2. why we need this proxy

jdneo replied:
Rename to `BspProxy` seems better

jdneo commented on extension/src/bs/JdtlsImporterConnector.ts:
```diff
@@ -0,0 +1,63 @@
+
+    private registerCommand(): void {
+        this.context.subscriptions.push(
+            vscode.commands.registerCommand(GET_IMPORTER_PIPE_NAME, (pipeName: string) => {
```
better naming: GET_IMPORTER_PIPE_NAME

jdneo commented on extension/src/bs/JdtlsImporterConnector.ts:
```diff
@@ -0,0 +1,63 @@
+        this.registerCommand();
+    }
+
+    //Receive the pipe name from Java jdt.ls importer
```
tsdoc above the method should use

```
/**
 * ...
 */
```

jdneo replied:
Please explain why we need to wait the message from importer.

Commit:
fix comments

jdneo commented on extension/src/bs/JdtlsImporterConnector.ts:
```diff
@@ -0,0 +1,72 @@
+import * as vscode from "vscode";
+import * as path from "path";
+
+export const ON_WILL_IMPORTER_READY = "gradle.onWillImporterReady";
```
onWillImporterConnect?

jdneo commented on extension/src/bs/BuildServerConnector.ts:
```diff
@@ -0,0 +1,37 @@
+import { generateRandomPipeName } from "../util/generateRandomPipeName";
+
+/**
+ * This class will create named pipe file and setting up a pipe server
```
creates named .....

jdneo commented on gradle-server/src/main/java/com/github/badsyntax/gradle/utils/Utils.java:
```diff
@@ -74,4 +76,34 @@ private static String toPackageName(String name) {
+		return paramMap;
+	}
+
+	public static int parseIntegerParam(Map<String, String> params, String key) throws IllegalArgumentException {
```
It's highly duplicated with `validateRequiredParam`

Just use `Integer.parseInt(validateRequiredParam(xxx))`;

Commit:
fix comments

jdneo committed:
Merge branch 'develop' into merge-with-named-pipe

Jiaaming committed:
Merge branch 'develop' into merge-with-named-pipe

Commit:
1. Add local build-server-for-gradle jar import and related gradle buildJar tasks. 2.Update debug logic. 3.fix typo and comments

Commit:
Fix: handle empty param in bspProxy repost

jdneo commented on extension/src/util/config.ts:
```diff
@@ -26,6 +35,72 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+        javaHome = getJavaHome() || undefined;
+        if (javaHome) {
+            javaVersion = await getMajorVersion(javaHome);
+            // Ensure the Java version is greater than 17
```
No need to add comment here. Because it's very clear that you are ensuring the java version is higher than 17. Besides, imagine that in the future, if the required version changes, you will need to update this comment as well.

I would suggest to adding comment at line 39, explaining where does the requirement comes from

jdneo commented on extension/src/util/config.ts:
```diff
@@ -26,6 +35,72 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+
+    // search valid JDKs from env.JAVA_HOME, env.PATH, SDKMAN, jEnv, jabba, Common directories
+    const javaRuntimes = await listJdks();
+    sortJdksByVersion(javaRuntimes);
```
Is sort required? Since you will scan the version of jdks in line4

jdneo commented on extension/src/util/config.ts:
```diff
@@ -26,6 +35,72 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+    if (validJdks.length > 0) {
+        sortJdksBySource(validJdks);
+        javaHome = validJdks[0].homedir;
+        javaVersion = validJdks[0].version!.major;
```
early return if found a valid jdk here

jdneo commented on gradle-server/build.gradle:
```diff
@@ -12,6 +12,9 @@ java {
 dependencies {
   implementation project(":gradle-plugin-api")
+  implementation files('build/libs/server.jar')
+  implementation fileTree(dir: 'build/libs/runtime', include: ['*.jar'], exclude: ['gradle-tooling-api-*.jar'])
```
Consider the transitive dependencies

jdneo commented on gradle-server/src/main/java/com/github/badsyntax/gradle/BuildServerThread.java:
```diff
@@ -0,0 +1,30 @@
+			String[] args = {"--pipe=" + this.pipeName};
+			Launcher.main(args);
+		} catch (Exception e) {
+			e.printStackTrace();
```
throw it or handle it.

Commit:
Fix: comments

Jiaaming committed:
Merge branch 'develop' into merge-with-named-pipe

jdneo commented on extension/build.gradle:
```diff
@@ -235,8 +235,18 @@ task copyBuildServerJars(type: Copy) {
   mustRunAfter copyDocs
 }
 
+task copyBuildServerJarsToGradleServer(type: Copy) {
```
Add a todo comment saying that this task can be removed once the build server is published.

jdneo commented on extension/src/Extension.ts:
```diff
@@ -312,7 +318,21 @@ export class Extension {
     private async restartServer(): Promise<void> {
         if (this.server.isReady()) {
```
Logic here may have some problem.

When the restartServer() is invoked, it may happen that builds are cancelled but the server is not restart.

I think you should let the restartServer() do a pure thing about restart. Which is said, ask if want to reload first, if yes, call the method `restartServer()`. Another benefit of doing this yes that we can send the info when settings are changed, to know how many users will change those settings during developing.

Jiaaming replied:
Yeah I think I should put `await this.client.cancelBuilds()` inside `if (selection === action)` statement.

jdneo commented on extension/src/server/GradleServer.ts:
```diff
@@ -29,51 +30,60 @@ export class GradleServer {
+        if (redHatJavaInstalled()) {
+            const javaExecPath = getRedHatJavaExecutablePath() || (await getJavaExecutablePath());
+            if (javaExecPath === undefined) {
                 await vscode.window.showErrorMessage(NO_JAVA_EXECUTABLE);
```
This will pops up a error message, is it by design?

Jiaaming replied:
Yes. Design here: 
if redHat.java installed: 
- Find redHat embeded jre --(if not found)--> try our best to find java executable with `jdk>=17`
- if still not found, **showErrorMessage**

If not installed:
- meaning that `build server`&`importer` won't be invoked, but we still need to move forward and start MergeServer

Merge Server will check if java executable exist and decide whether start `build server` or not.

Commit:
1.Fix: comment

2.Refine debug setting logic

3.Rename: gradle service -> task server for better understanding

Commit:
1.Add error handle for rpc connection

2.Add telemetry

Commit:
Fix: typo

Jiaaming commented on gradle-server/build.gradle:
Since we need `build-server-for-gradle` to be dependency, I think we also need to update `gradle-server`'s Java version to be 17. After we merge `GradleLangangeServer`, the whole project should use jdk>=17
```
java {
  sourceCompatibility = JavaVersion.VERSION_1_8
  targetCompatibility = JavaVersion.VERSION_1_8
}
```

jdneo replied:
Correct!

Jiaaming commented:
## Restart Summaries and Solutions
All these cases require user to response to the reload/restart window.
### 1. Gradle (Task) Server Stop by accident and Process Exit
- **Previous Solution:** 
  - Restart the Gradle (Task) Server by calling `this.start()` in `GradleServer.ts`.
  - Code reference: [GradleServer.ts#L70-L72](https://github.com/microsoft/vscode-gradle/blob/d904c27d96595d9f83aff251d85072502019559b/extension/src/server/GradleServer.ts#L70-L72)
- **Current Approach:** 
  - Reload the window.
- **Note:** 
  - Reloading is necessary after we merge

### 2. Gradle (Task) Client Loses Connection
- **Previous Solution:** 
  - Reconnect to the Gradle (Task) Server in `GradleClient.ts`.
  - Code reference: [GradleClient.ts#L491-L500](https://github.com/microsoft/vscode-gradle/blob/d904c27d96595d9f83aff251d85072502019559b/extension/src/client/GradleClient.ts#L491-L500)
- **Current Approach:** 
  - Same as before or Reload the window, depending on the situation.
- **Note:** 
  - Typically occurs alongside server downtime. Actions here may not make a difference if the server itself requires a reload. Need more discussion here.

### 3. User Changes Configuration
- Will change if we modify:
   1. `java.home || java.jdt.ls.java.home || java.import.gradle.java.home`
   2. project's Gradle wrapper properties file
- **Previous Solution:** 
  - Call `GradleServer.restart()`, which sets `this.restarting = true`, kills the current process, and restarts the server.
  - Code reference: [Extension.ts#L312-L317](https://github.com/microsoft/vscode-gradle/blob/d904c27d96595d9f83aff251d85072502019559b/extension/src/Extension.ts#L312-L317)
- **Current Approach:** 
  - Reload the window.
- **Note:** 
  - Necessary to apply configuration changes effectively, requiring a window reload.

Commit:
1.Update github workflow to feat local dependency

 2.Update GraleClient restart logic

3.Update gradle-server jdk version>=17

Commit:
Update: lint ignore

Commit:
fix: powermock jdk17

Jiaaming commented on extension/src/bs/BspProxy.ts:
```diff
@@ -0,0 +1,95 @@
+            }
+            importerConnection?.sendNotification(method);
+        });
+        importerConnection?.onError(([error, message, code]) => {
+            this.logger.error(`Error on importerConnection: ${error.message}`);
+            sendInfo("", {
+                kind: "bspProxy-importerConnectionError",
+                data2: JSON.stringify({
+                    error,
+                    message,
+                    code,
+                }),
+            });
+            // TODO: Implement more specific error handling logic here
+        });
+
+        buildServerConnection?.onError(([error, message, code]) => {
+            this.logger.error(`Error on buildServerConnection: ${error.message}`);
+            sendInfo("", {
+                kind: "bspProxy-importerConnectionError",
+                data2: JSON.stringify({
+                    error,
+                    message,
+                    code,
+                }),
+            });
+            // TODO: Implement more specific error handling logic here
+        });
```
`rpc.MessageConnection` provide an `onError` listerner. I'm not sure when & how this connection might occur error, so I think we can collect Telemetry here to get more insight

jdneo committed:
Merge branch 'develop' into merge-with-named-pipe

jdneo commented on CONTRIBUTING.md:
```diff
@@ -40,34 +40,18 @@ The extension uses a Gradle plugin (`com.microsoft.gradle.GradlePlugin`) to get
+1. Run vscode launch configuration `Debug Gradle Server & Extension`.
+2. Run vscode launch configuration `Attach to Gradle Server` when you notice the `Gradle: Connecting...` message in the bottom status bar.
+
+> Note: If `Java: Error` message appear in the bottom status bar, it indicates that the connection attempt in step 2 was too slow. [GradleBuildClient](/extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/ImporterPlugin.java#L107) requires the Gradle Server to be active to establish a connection. If this happens, retry the attachment more swiftly.
```
Can we provide a sample of the specific error?

jdneo commented on extension/src/Extension.ts:
```diff
@@ -312,10 +333,17 @@ export class Extension {
     private async restartServer(): Promise<void> {
         if (this.server.isReady()) {
```
If the server is not ready, does that mean nothing will happen here?

Jiaaming replied:
`this.server.isReady()` will be `false` if the process exit with `code!=0`, which will be catch by another Telemetry and also require user to restart(reload), so I think here would be fine.

jdneo commented on extension/src/server/GradleServer.ts:
```diff
@@ -29,65 +32,84 @@ export class GradleServer {
+        }
+        let startBuildServer = "false";
+        if (redHatJavaInstalled()) {
+            const javaExecPath = getRedHatJavaExecutablePath() || (await getJavaExecutablePath());
```
Since now both servers are on jdk 17, I think the logic here can be simplified

Jiaaming commented on extension/src/server/GradleServer.ts:
```diff
@@ -29,65 +32,84 @@ export class GradleServer {
-            this.fireOnStart();
         }
+        const buildServerPipeName = this.bspProxy.getBuildServerPipeName();
+        const args = [
+            `--port=${this.taskServerPort}`,
+            `--pipeName=${buildServerPipeName}`,
+            `--bundleDir=${bundleDirectory}`,
+            `--startBuildServer=${startBuildServer}`,
+        ];
```
It don't need to send `bundleDirectory` and `buildServerPipeName` if `startBuildServer == false`, will fix it

Commit:
Refine jdk search logic that start gradle-server

Jiaaming commented on extension/src/server/serverUtil.ts:
```diff
@@ -16,7 +16,7 @@ export interface ProcessEnv {
 
 export async function getGradleServerEnv(): Promise<ProcessEnv | undefined> {
-    const javaHome = await getSupportedJavaHome();
+    const javaHome = getRedHatJavaEmbeddedJRE() || (await findValidJavaHome());
```
The `javaHome` name here might be a bit confuse here.
 `getGradleServerEnv()` will add `VSCODE_JAVA_HOME` env variable here, which will be used by the [GradleServer StartScript](https://github.com/microsoft/vscode-gradle/blob/a7584c3fbbf0915bc537e8c8ea6802bfeb5f33ce/gradle-server/src/main/resources/startScriptTemplates/unixStartScript.txt#L86-L97)
In this script, the purpose here is to get a Java executable path, so the `redHat embedded jre` path here would be acceptable.

jdneo commented on gradle-server/src/main/java/com/github/badsyntax/gradle/GradleServer.java:
```diff
@@ -42,24 +44,48 @@ public void run() {
+				server.start();
+				server.blockUntilShutdown();
+			} catch (IOException | InterruptedException e) {
+				e.printStackTrace();
```
throw

jdneo commented on gradle-server/src/main/java/com/github/badsyntax/gradle/GradleServer.java:
```diff
@@ -42,24 +44,48 @@ public void run() {
+		int taskServerPort = Integer.parseInt(Utils.validateRequiredParam(params, "port"));
+		boolean startBuildServer = Boolean.parseBoolean(Utils.validateRequiredParam(params, "startBuildServer"));
+
+		startTaskServerThread(taskServerPort);
```
Move this line after `int taskServerPort = Integer.parseInt(Utils.validateRequiredParam(params, "port"));`

jdneo commented on extension/src/Extension.ts:
```diff
@@ -310,10 +331,15 @@ export class Extension {
+    }
+
+    private async showRestartWindow(): Promise<string | undefined> {
+        const msg = "Please reload the window to make the change take effect. Reload now?";
```
What if we use `workbench.action.restartExtensionHost
`

Commit:
Fix: comments

jdneo commented on extension/src/util/config.ts:
```diff
@@ -26,17 +28,59 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+}
+
+export async function findValidJavaHome(): Promise<string | undefined> {
+    const REQUIRED_JDK_VERSION = 17;
```
This const can be out of this function's scope

Commit:
Fix: comment

jdneo commented on extension/src/Extension.ts:
```diff
@@ -328,7 +354,19 @@ export class Extension {
-                    await this.restartServer();
+                    const selection = await this.showRestartWindow();
+                    if (selection === "Reload") {
+                        sendInfo("", {
```
Can we merge the two `sendInfo` into one?

jdneo commented on extension/src/Extension.ts:
```diff
@@ -301,7 +310,19 @@ export class Extension {
-                await this.restartServer();
+                const selection = await this.showRestartWindow();
+                if (selection === "Reload") {
+                    sendInfo("", {
```
Can we merge the two `sendInfo` into one?

jdneo commented on extension/src/server/GradleServer.ts:
```diff
@@ -29,65 +32,74 @@ export class GradleServer {
         );
         if (input === OPT_RESTART) {
-            await this.start();
+            sendInfo("", {
```
Can we merge the two `sendInfo` into one?

jdneo commented on extension/src/Extension.ts:
```diff
@@ -328,7 +354,19 @@ export class Extension {
                 ) {
-                    await this.restartServer();
+                    const selection = await this.showRestartWindow();
+                    if (selection === "Reload") {
```
wording

jdneo commented on extension/src/bs/BspProxy.ts:
```diff
@@ -0,0 +1,95 @@
+            this.logger.error(`Error on importerConnection: ${error.message}`);
+            sendInfo("", {
+                kind: "bspProxy-importerConnectionError",
+                data2: JSON.stringify({
```
`error` -> `errorStack`
`message` -> `message`

Commit:
fix comments

jdneo commented on extension/src/Extension.ts:
```diff
@@ -38,6 +38,7 @@ import { GradleBuildContentProvider } from "./client/GradleBuildContentProvider"
 import { BuildServerController } from "./bs/BuildServerController";
 import { GradleTestRunner } from "./bs/GradleTestRunner";
 import { BspProxy } from "./bs/BspProxy";
+const OPT_RESTART = "Restart";
```
Move to `src/constant.ts`

Commit:
Fix: comment

jdneo commented on extension/build.gradle:
```diff
@@ -235,8 +235,19 @@ task copyBuildServerJars(type: Copy) {
+
 task buildJars() {
-  dependsOn copyJdtlsPluginJar, copyBuildServerJars
+  dependsOn copyJdtlsPluginJar, copyBuildServerJars, copyBuildServerJarsToGradleServer
```
Is `copyBuildServerJars` still needed?

jdneo commented on extension/build.gradle:
```diff
@@ -235,8 +235,19 @@ task copyBuildServerJars(type: Copy) {
+    include '**/*.jar'
+    include '**/init.gradle'
+  }
+  into '../gradle-server/build/libs/'
```
If directly copying those jars into `gradle-server/build/libs/`, Then the exclude makes no sense

jdneo commented on extension/src/util/config.ts:
```diff
@@ -26,17 +29,58 @@ export function getConfigJavaImportGradleJavaHome(): string | null {
+
+    // Search valid JDKs from env.JAVA_HOME, env.PATH, SDKMAN, jEnv, jabba, common directories
+    const javaRuntimes = await listJdks();
+    const validJdks = javaRuntimes.filter((r) => r.version!.major >= REQUIRED_JDK_VERSION);
```
Use `.find()` to speed up. since you only need the first element meets the requirement

Commit:
Fix: build.gradle file

Commit:
Fix: extension start sequence

jdneo reviewed (approved):
LGTM
